include/ring_buffer.h

```diff
@@ -242,6 +287,142 @@ __deprecated static inline int sys_ring_buf_get(struct ring_buf *buf,
+ *
+ * @param[in]  buf  Address of ring buffer.
+ * @param[out] data Pointer to the address. It is set to a location within
+ *		    ring buffer.
+ * @param[in]  size Requested allocation size (in bytes).
+ *
+ * @return Size of allocated buffer which can be smaller than requested if
+ *	   there is not enough free space or buffer wraps.
+ */
+u32_t ring_buf_put_claim(struct ring_buf *buf, u8_t **data, u32_t size);
+
+/**
+ * @brief Indicate number of bytes written to allocated buffers.
+ *
+ * @warning
+ * Use cases involving multiple writers to the ring buffer must prevent
+ * concurrent write operations, either by preventing all writers from
+ * being preempted or by using a mutex to govern writes to the ring buffer.
+ *
+ * @warning
+ * Ring buffer instance should not mix byte access and item access
+ * (calls prefixed with ring_buf_item_).
+ *
+ * @param  buf  Address of ring buffer.
+ * @param  size Number of valid bytes in the allocated buffers.
+ *
+ * @retval 0 Successful operation.
+ * @retval -EINVAL Provided @a size exceeds free space in the ring buffer.
+ */
+int ring_buf_put_finish(struct ring_buf *buf, u32_t size);
+
+/**
+ * @brief Write (copy) data to a ring buffer.
+ *
+ * This routine writes data to a ring buffer @a buf.
+ *
+ * @warning
+ * Use cases involving multiple writers to the ring buffer must prevent
+ * concurrent write operations, either by preventing all writers from
+ * being preempted or by using a mutex to govern writes to the ring buffer.
+ *
+ * @warning
+ * Ring buffer instance should not mix byte access and item access
+ * (calls prefixed with ring_buf_item_).
+ *
+ * @param buf Address of ring buffer.
+ * @param data Address of data.
+ * @param size Data size (in bytes).
+ *
+ * @retval Number of bytes written.
+ */
+u32_t ring_buf_put(struct ring_buf *buf, const u8_t *data, u32_t size);
+
+/**
+ * @brief Get address of a valid data in a ring buffer.
+ *
+ * With this routine, memory copying can be reduced since internal ring buffer
+ * can be used directly by the user. Once data is processed it can be freed
+ * using @ref ring_buf_get_finish.
+ *
+ * @warning
+ * Use cases involving multiple reads of the ring buffer must prevent
+ * concurrent read operations, either by preventing all readers from
+ * being preempted or by using a mutex to govern reads to the ring buffer.
+ *
+ * @warning
+ * Ring buffer instance should not mix byte access and item access
+ * (calls prefixed with ring_buf_item_).
+ *
+ * @param[in]  buf  Address of ring buffer.
+ * @param[out] data Pointer to the address. It is set to a location within
+ *		    ring buffer.
+ * @param[in]  size Requested size (in bytes).
+ *
+ * @return Number of valid bytes in the provided buffer which can be smaller
+ *	   than requested if there is not enough free space or buffer wraps.
+ */
+u32_t ring_buf_get_claim(struct ring_buf *buf, u8_t **data, u32_t size);
+
+/**
+ * @brief Indicate number of bytes that can be freed.
```

Comment:
IMHO this description is a bit weak and unclear.

Compare it with description for put_finish: 

> Indicate number of bytes written to allocated buffers.


Would be nice if it also talk about bytes read, etc.